### ISSUE TITLE
Add support of delete operation to Redis Sink

### ISSUE THREAD

**Issue by isaranchuk** [lensesio/stream-reactor#640]
[Redis Sink](https://docs.confluent.io/current/connect/kafka-connect-redis/index.html) connector from Confluent supports delete operation.
If the record stored in Kafka has a null value, this connector sends a delete with the corresponding key to Redis.

I'm wondering if there're any plans to add support of delete operation to the current Redis Sink connector?

**Comment by andrewstevenson** [comment]
@isaranchuk I haven't looked at confluents sink, I assume it there is a one to one mapping between the key in the Kafka message and Redis, this should be easy to add. Do you fancy contributing a PR?

You would need to add jedis.del() and jedis.zrem() to the writer classes in here https://github.com/lensesio/stream-reactor/tree/master/kafka-connect-redis/src/main/scala/com/datamountaineer/streamreactor/connect/redis/sink/writer

**Comment by isaranchuk** [comment]
@andrewstevenson yes from what I see it's a one to one mapping in the confluent connector.

At the moment I'm checking different options and if deletion on connector side is required I will prepare a PR for that.

**Comment by hsomyk** [comment]
@andrewstevenson @isaranchuk 

I suggest to implement DELETE functionality in the same way as it is implemented in Cassandra Sink Connector.

For example when Redis writer receive record with null record value it should delete this item from Redis. In order to build key for DELETE operation each Redis writer should use the same KCQL configurations as used for the INSERT operations the only difference is that for composing key for DELETE commands we will use received record key data instead of using record value data.
It should skip the record in case of composing key failure.

Probably it is good idea also to add some configuration flag/switcher to turn on/off this delete functionality.

@andrewstevenson, so what do you think about this way of implementing Redis Delete functionality?

**Comment by andrewstevenson** [comment]
@isaranchuk @hsomyk i have it in a branch. I just need to write some tests. I'll push tomorrow

**Comment by isaranchuk** [comment]
@andrewstevenson did you have a chance to push your changes so we can have a look? :smiley:

**Comment by andrewstevenson** [comment]
@isaranchuk I've pushed to feat/redis-delete but I don't have unit tests yet. If you want to take over this ticket feel free, we should also bump the Jedis version so we can bring in support for Redis Streams. Fancy giving this a go?

**Comment by isaranchuk** [comment]
@andrewstevenson thanks!
I'm wondering if we can implement some kind of a delete handler where we have access to Redis connection, so that delete handler is invoked instead of a default mechanism if needed?

I'm asking because our use case is not that simple as we don't have a one to one mapping between Kafka and Redis keys.
As a result, we have to implement our own logic for deletion.

**Comment by andrewstevenson** [comment]
@isaranchuk Its possible to create interface. The default would be the one to one mapping I put in but the user can supply a custom delete class for it.  We do something similar in JMS, MQTT and FTP

**Comment by isaranchuk** [comment]
@andrewstevenson could you please share a link where you have the support of such customization?

**Comment by andrewstevenson** [comment]
@isaranchuk Theres an example [here](https://github.com/lensesio/stream-reactor/blob/master/kafka-connect-jms/src/main/scala/com/datamountaineer/streamreactor/connect/jms/config/JMSSettings.scala#L121). You would need to introduce a delete trait that has a function that accepts the SinkRecord inside which you do want ever you want and returns the keys to delete from Redis. Then you allow the use to set the class they want to use that implements this trait.

**Comment by isaranchuk** [comment]
@andrewstevenson not sure that the list of Redis keys is enough as these keys can use different delete operations, e.g. DEL, ZREM, etc.

What about having a delete trait that has access to `jedis` so we can have our custom delete functionality?

**Comment by andrewstevenson** [comment]
@isaranchuk There's a writer implementing the behaviour already for each Redis structure we support. I'm not a Redis expert but are you saying that each structure has one or more ways to delete i.e. a sortedSet can use both `del` and `rem`?

If not then in each writer we call the delete implementation to get the keys?

**Comment by isaranchuk** [comment]
@andrewstevenson yes, Redis has one way to delete for each structure, so no problems with that.

Let me explain my use case in more detail to have better understanding why I'm asking about access to`jedis` in the delete handler.

We don't have a simple mapping between Kafka and Redis keys, because we build Redis key based on properties that are present in Kafka value using Kafka Connect transformers.

In the case of delete, Kafka value will be null, so there're no chance to build keys properly.
On the other side we can use Kafka key to implement some kind of a "smart" deletion, but we need Redis connection for that.

**Comment by isaranchuk** [comment]
@andrewstevenson yes, Redis has one way to delete for each structure, so no problems with that.

Let me explain my use case in more detail to have better understanding why I'm asking about access to `jedis` in the delete handler.

We don't have a simple mapping between Kafka and Redis keys, because we build Redis key based on properties that are present in Kafka value using Kafka Connect transformers.

In the case of delete, Kafka value will be null, so now there's no chance to build keys properly.
On the other side, we can use Kafka key to implement some kind of a "smart" deletion, but we need Redis connection for that.

**Comment by andrewstevenson** [comment]
@isaranchuk if you look at the writer classes, one for each redis structure it has access to jedis. If you write a delete handler you just need return the key to delete. We don't need to pass in jedis, just the sink record, which includes the kafka keys. You can then contruct the redis key to delete from the SinkRecord and return that for the writer to delete.  I think we have a key builder when doing the insert so the logic is the same. Pass in a SinkRecord, which contains the kafka key and value and get a string back representing the redis key.

**Comment by isaranchuk** [comment]
@andrewstevenson the problem is that we can't build keys for deletion based on SinkRecord.
In our case, we have to query Redis first and then do the deletion.

**Comment by andrewstevenson** [comment]
@isaranchuk fair enough.

**Comment by isaranchuk** [comment]
@andrewstevenson I understand that our requirement is very unique, so maybe we should proceed with our custom connector for that.

But I'm wondering if we can introduce design change where we can override default delete handler with the custom one that has access to Redis connection?
In that case, custom delete handler should be used per connector not per writer.

**Comment by andrewstevenson** [comment]
@isaranchuk You would just have a trait that has a function. I think you want it per writer as one connector can handle multiple topics (KCQL stmts) so you might have different requirements per topic. Now it might be better to pass in the jedis once.

def delete(record: SinkRecord, jedis: Jedis) {
//custom code
}

**Comment by isaranchuk** [comment]
@andrewstevenson you're completely right that connector can handle multiple topics.
But here we're talking about sink connector and not a source connector, so it doesn't matter how many topics are there as we just need to delete from Redis.

I do agree that default delete functionality must be per writer.
But in case of a custom delete handler, we can have it per connector, so it can process Kafka record one time if we have many KCQL queries configured.

**Comment by andrewstevenson** [comment]
@isaranchuk We don't have a Redis source. My point is that there's no reason to enforce a connector wide handler when we can handle it at the writer level and give more flexibility and the ability to have custom deletion per topic. We have clients using the connector who have multiple KCQL stmts in one connector with different topics and payloads.

we can then use kcql and do something like

> insert into redis_cache_1 select * from topic1 withconverter my.custom.converter

**Comment by doracl** [comment]
+1
